File: SinglePointBatch/scan.f90
!Generate a scan along a internal coordinate
!Input : internal coordinate definition, geom
!Output: geom.data
!Dependency: my Fortran-Library, as written in makefile
program main
    use FortranLibrary
    implicit none
!Molecule information
    integer::NAtoms,intdim,cartdim
    character*2,allocatable,dimension(:)::ElementSymbol
    real*8,allocatable,dimension(:)::ElementNumber,mass,r0,q0
!Work variable
    character*32::chartemp; logical::flag; integer::i,j
    real*8,allocatable,dimension(:)::q,r
!Initialize
    inquire(file='InternalCoordinateDefinition',exist=flag)
    if(flag) then
        chartemp=''
        call DefineInternalCoordinate(chartemp,intdim)
    else
        chartemp='Columbus7'
        call DefineInternalCoordinate(chartemp,intdim)
    end if
    open(unit=99,file='geom',status='old')!Read reference geometry
        NAtoms=0; do; read(99,*,iostat=i); if(i/=0) exit; NAtoms=NAtoms+1; end do; rewind 99
        allocate(ElementSymbol(NAtoms)); allocate(ElementNumber(NAtoms))
        allocate(r0(3*NAtoms)); allocate(mass(NAtoms))
        do i=1,NAtoms
            read(99,*)ElementSymbol(i),ElementNumber(i),r0(3*i-2:3*i),mass(i)
        end do
        mass=mass*AMUInAU
    close(99)
    cartdim=3*NAtoms
    allocate(q0(intdim)); q0=InternalCoordinateq(r0,intdim,cartdim)
!Generate a scan along a internal coordinate
    chartemp='assimilate'
    allocate(q(intdim)); allocate(r(cartdim))
    open(unit=99,file='geom.data',status='replace')
        !Example: start from q0, scan 1st internal coordinate
        do i=-5,5
            q=q0
            q(1)=q(1)+dble(i)*0.01d0
            r=CartesianCoordinater(q,cartdim,intdim,uniquify=chartemp,mass=mass,r0=r0)
            do j=1,NAtoms
                write(99,'(1x,A2,2x,F5.1,4F14.8)')ElementSymbol(j),ElementNumber(j),r(3*j-2:3*j),mass(j)/AMUInAU
            end do
        end do
    close(99)
end program main

File: SinglePointBatch/scan.py
"""
Generate a scan along an internal coordinate
"""

''' Library '''
import argparse
from pathlib import Path
import shutil
import numpy
import FortranLibrary as FL
import basic

''' Routine '''
def parse_args() -> argparse.Namespace: # Command line input
    parser = argparse.ArgumentParser(__doc__)
    parser.add_argument('IntCDefFormat', type=str, help='internal coordinate definition format: Columbus7 or default')
    parser.add_argument('IntCDef', type=Path, help='internal coordinate definition file')
    parser.add_argument('geom',    type=Path, help='Columbus7 geometry file')
    parser.add_argument('IntC2scan', type=int, help='internal coordinate to scan')
    parser.add_argument('-b','--bidirection', action='store_true', help='scan bidirectionsally (default = positive only)')
    parser.add_argument('-n','--NSteps', type=int, default=10  , help='number of scan steps (default = 10)')
    parser.add_argument('-l','--length', type=int, default=0.01, help='step length (default = 0.01)')
    parser.add_argument('-o','--output', type=Path, default='geom.data', help='output file (default = geom.data), will append if already exists')
    args = parser.parse_args()
    return args

if __name__ == "__main__":
    ''' Initialize '''
    args = parse_args()
    # Define internal coordinate
    if args.IntCDefFormat == 'Columbus7': IntCDef = Path('intcfl') 
    else: IntCDef = Path('InternalCoordinateDefinition')
    if args.IntCDef != IntCDef:
        shutil.copy(args.IntCDef, IntCDef)
        intdim, intcdef = FL.FetchInternalCoordinateDefinition(args.IntCDefFormat)
        FL.DefineInternalCoordinate('Columbus7')
        IntCDef.unlink()
    else:
        intdim, intcdef = FL.FetchInternalCoordinateDefinition(args.IntCDefFormat)
        FL.DefineInternalCoordinate('Columbus7')
    # Read geometry
    NAtoms, symbol, number, r, mass = basic.read_geom(args.geom)
    cartdim = 3 * NAtoms
    q = numpy.empty(intdim)
    FL.InternalCoordinateq(r, q, cartdim, intdim)
    ''' Do the job '''
    q1 = numpy.empty(q.shape); r1 = numpy.empty(r.shape)
    if args.bidirection:
        for i in range(-args.NSteps, 0):
            q1[:] = q[:]; q1[args.IntC2scan-1] += i * args.length
            FL.CartesianCoordinater(q1, r1, intdim, cartdim, uniquify='assimilate', mass=mass, r0=r)
    for i in range(1, args.NSteps+1):
        q1[:] = q[:]; q1[args.IntC2scan-1] += i * args.length
        FL.CartesianCoordinater(q1, r1, intdim, cartdim, uniquify='assimilate', mass=mass, r0=r)
        basic.write_geom(args.output, NAtoms, symbol, number, r, mass)